File: SmartMedical-Angular/src/app/core/services/appointment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Appointment {
  id?: string;
  patientId?: string;
  patientName?: string;
  patientEmail?: string;
  patientPhone?: string;
  doctorId?: string;
  doctorName?: string;
  doctorSpecialty?: string;
  appointmentDate?: string;
  appointmentTime?: string;
  reason?: string;
  status?: string;
  createdAt?: string;
}

export interface AppointmentRequest {
  patientId?: string;
  patientName: string;
  patientEmail: string;
  patientPhone: string;
  doctorId: string;
  appointmentDate: string;
  appointmentTime: string;
  reason: string;
}

@Injectable({ providedIn: 'root' })
export class AppointmentService {
  private apiUrl = `${environment.gatewayUrl}/appointments`;

  constructor(private http: HttpClient) {}

  create(request: AppointmentRequest): Observable<any> {
    return this.http.post(this.apiUrl, request);
  }

  getAll(): Observable<Appointment[]> {
    return this.http.get<Appointment[]>(this.apiUrl);
  }

  getByPatientEmail(email: string): Observable<Appointment[]> {
    return this.http.get<Appointment[]>(`${this.apiUrl}/patient/${encodeURIComponent(email)}`);
  }

  updateStatus(id: string, status: string): Observable<Appointment> {
    return this.http.patch<Appointment>(`${this.apiUrl}/${id}/status`, { status });
  }

  delete(id: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${id}`);
  }
}

File: Backend/AUTH-SERVICE/src/main/java/tn/esprit/authservice/service/AppointmentService.java
package tn.esprit.authservice.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import tn.esprit.authservice.dto.AppointmentRequest;
import tn.esprit.authservice.entity.Appointment;
import tn.esprit.authservice.entity.Doctor;
import tn.esprit.authservice.repository.AppointmentRepository;
import tn.esprit.authservice.repository.DoctorRepository;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class AppointmentService {

    private final AppointmentRepository appointmentRepository;
    private final DoctorRepository doctorRepository;

    public Appointment create(AppointmentRequest request) {
        if (request.getDoctorId() == null || request.getDoctorId().isBlank()) {
            throw new RuntimeException("Médecin requis");
        }
        if (request.getAppointmentDate() == null || request.getAppointmentTime() == null) {
            throw new RuntimeException("Date et heure requises");
        }

        Doctor doctor = doctorRepository.findById(request.getDoctorId())
                .orElseThrow(() -> new RuntimeException("Médecin non trouvé"));

        if (!doctor.isActive()) {
            throw new RuntimeException("Ce médecin n'accepte pas de nouveaux rendez-vous");
        }

        Appointment appointment = Appointment.builder()
                .patientId(request.getPatientId())
                .patientName(request.getPatientName())
                .patientEmail(request.getPatientEmail())
                .patientPhone(request.getPatientPhone())
                .doctorId(doctor.getId())
                .doctorName(doctor.getFirstName() + " " + doctor.getLastName())
                .doctorSpecialty(doctor.getSpecialization())
                .appointmentDate(request.getAppointmentDate())
                .appointmentTime(request.getAppointmentTime())
                .reason(request.getReason())
                .status("PENDING")
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        return appointmentRepository.save(appointment);
    }

    public List<Appointment> getAll() {
        return appointmentRepository.findAllByOrderByCreatedAtDesc();
    }

    public List<Appointment> getByPatientEmail(String email) {
        return appointmentRepository.findByPatientEmailOrderByCreatedAtDesc(email);
    }

    public List<Appointment> getByDoctorId(String doctorId) {
        return appointmentRepository.findByDoctorIdOrderByCreatedAtDesc(doctorId);
    }

    public Appointment updateStatus(String id, String status) {
        Appointment appointment = appointmentRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Rendez-vous non trouvé"));
        appointment.setStatus(status);
        appointment.setUpdatedAt(LocalDateTime.now());
        return appointmentRepository.save(appointment);
    }

    public void delete(String id) {
        appointmentRepository.deleteById(id);
    }
}
